src/node/transaction.cpp

```diff
@@ -36,10 +36,28 @@ TransactionError BroadcastTransaction(NodeContext& node, const CTransactionRef t
         if (!existingCoin.IsSpent()) return TransactionError::ALREADY_IN_CHAIN;
     }
     if (!node.mempool->exists(hashTx)) {
-        // Transaction is not already in the mempool. Submit it.
+        // Transaction is not already in the mempool.
         TxValidationState state;
+        CAmount fee{0};
+        if (max_tx_fee) {
+            if (!GetTransactionFee(*node.mempool, state, tx, &fee)) {
+                err_string = state.ToString();
+                if (state.IsInvalid()) {
+                    if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
+                        return TransactionError::MISSING_INPUTS;
+                    }
+                    return TransactionError::MEMPOOL_REJECTED;
+                } else {
+                    return TransactionError::MEMPOOL_ERROR;
+                }
+            }
+            else if (fee > max_tx_fee) {
+                return TransactionError::MAX_FEE_EXCEEDED;
+            }
+        }
+        // Transaction fee is acceptable. Submit the transaction.
         if (!AcceptToMemoryPool(*node.mempool, state, tx,
-                                nullptr /* plTxnReplaced */, false /* bypass_limits */, max_tx_fee)) {
+                nullptr /* plTxnReplaced */, false /* bypass_limits */)) {
```

Comment:
Similar nonblocking suggestion as above -- You may even want to consider refactoring this common error processing into a `TransactionError` method.
```cpp
        if (!AcceptToMemoryPool(*node.mempool, state, tx,
                nullptr /* plTxnReplaced */, false /* bypass_limits */)) {
            err_string = state.ToString();
            if (!state.IsInvalid()) {
                return TransactionError::MEMPOOL_ERROR;
            }
            if (state.GetResult() == TxValidationResult::TX_MISSING_INPUTS) {
                return TransactionError::MISSING_INPUTS;
            }
            return TransactionError::MEMPOOL_REJECTED;
        }
```

Reply:
Good point, made a helper function :)